lib/std/event/lock.zig

```diff
@@ -16,107 +16,90 @@ const Loop = std.event.Loop;
-    };
 
-    pub fn init() Lock {
-        return Lock{
-            .shared = false,
-            .queue = Queue.init(),
-            .queue_empty = true,
-        };
-    }
+        var waiter: Waiter = undefined;
+        waiter.next = null;
+        waiter.tail = &waiter; 
 
-    pub fn initLocked() Lock {
-        return Lock{
-            .shared = true,
-            .queue = Queue.init(),
-            .queue_empty = true,
+        const head = switch (self.head) {
+            UNLOCKED => unreachable,
+            LOCKED => null,
+            else => @intToPtr(?*Waiter, self.head),
         };
-    }
-
-    /// Must be called when not locked. Not thread safe.
-    /// All calls to acquire() and release() must complete before calling deinit().
-    pub fn deinit(self: *Lock) void {
-        assert(!self.shared);
-        while (self.queue.get()) |node| resume node.data;
-    }
 
-    pub fn acquire(self: *Lock) callconv(.Async) Held {
-        var my_tick_node = Loop.NextTickNode.init(@frame());
+        if (head) |h| {
+            h.tail.next = &waiter;
+            h.tail = &waiter;
+        } else {
+            self.head = @ptrToInt(&waiter);
+        }
 
-        errdefer _ = self.queue.remove(&my_tick_node); // TODO test canceling an acquire
         suspend {
-            self.queue.put(&my_tick_node);
-
-            // At this point, we are in the queue, so we might have already been resumed.
+            waiter.node = Loop.NextTickNode{
+                .prev = undefined,
+                .next = undefined,
+                .data = @frame(),
+            };
+            held.release();
+        }
 
-            // We set this bit so that later we can rely on the fact, that if queue_empty == true, some actor
-            // will attempt to grab the lock.
-            @atomicStore(bool, &self.queue_empty, false, .SeqCst);
+        return Held{ .lock = self };
+    }
 
-            if (!@atomicRmw(bool, &self.shared, .Xchg, true, .SeqCst)) {
-                if (self.queue.get()) |node| {
-                    // Whether this node is us or someone else, we tail resume it.
-                    resume node.data;
+    pub const Held = struct {
+        lock: *Lock, 
+    
+        pub fn release(self: Held) void {
+            const waiter = blk: {
+                const held = self.lock.mutex.acquire();
+                defer held.release();
+
+                switch (self.lock.head) {
+                    UNLOCKED => {
+                        std.debug.panic("Lock unlocked when already unlocked", .{});
+                    },
+                    LOCKED => {
+                        self.lock.head = UNLOCKED;
+                        break :blk null;
+                    },
+                    else => {
+                        const waiter = @intToPtr(*Waiter, self.lock.head);
+                        self.lock.head = if (waiter.next == null) LOCKED else @ptrToInt(waiter.next);
+                        if (waiter.next) |next|
```

Comment:
should this be `|*next|`?

Reply:
the next pointer is read then dereferenced (i.e. `waiter.next.?.tail = waiter.tail;`) instead of mutating the `waiter.next` field itself.